### ISSUE TITLE
Parameter Server Proposal - Async, Streaming

### ISSUE THREAD

**Issue by crockpotveggies** [deeplearning4j/deeplearning4j#1601]
As has been discussed in the Gitter channels and offline by various contributors, a parameter server is being considered to significantly speed up distributed training times and remove bottlenecks. To encourage development and move this along, I've researched best practices from the literature and assembled them into a proposal.

I'll be pasting revisions below. I'd like to use this issue as a thread for discussion and implementation.

> [Streaming Async DL4J.docx](https://github.com/deeplearning4j/deeplearning4j/files/285410/Streaming.Async.DL4J.docx)

A visual from the proposal:
![dl4j distributed training](https://cloud.githubusercontent.com/assets/1445295/15591396/be81afa4-2351-11e6-91fc-4aceb545587e.png)

Feedback, critique, and questions are welcomed.

**Comment by jpatanooga** [comment]
some thoughts:

1. parameter servers are nice in design, but can be complex in practice
2. Adam and I have sat down with Alex Smola and talked about different ways
to do this (I'd look to his work for ideas)
3. has anyone hit the parameter size limitations of basic parameter
averaging? (probably no)
4. adding more daemons to a spark job is non-trivial work and DL4J is
mostly focused on production integrations w spark on hadoop distros. This
complicates factors.
5. The bottleneck effect can occur --- but the reason Vorpal Wabbit is
synchronous (per a f2f conversation w John Langford) is that its really
hard to debug async parallel iterative SGD
6. Again: finding a way to do this cleanly on spark is the key here. Adding
more systems to support another daemon / server is non-trivial from a
dev-Ops standpoint and hurts the distro in enterprise production situation.

Not trying to discourage this thread yet want to lay in some history and
context. Parameter servers end up being many times an engineering exercise.
I'd find a use case that demands the functionality of a parameter server to
start with, and then find a way to make it easily work with spark w/o
adding custom / extra dependencies. Doing those two things would greatly
increase the drive towards realizing a solid addition of a parameter server
to DL4J.


Older presentations we've done as part of the study of parallel iterative
algorithms:

http://www.slideshare.net/jpatanooga/hadoop-summit-2014-san-jose-introduction-to-deep-learning-on-hadoop

The roots of the current DL4J parallel design come from an older project
called "Metronome"
http://www.slideshare.net/jpatanooga/metronome-ml-confnov2013v20131113

(hadoop world strata 2012 talk) -- and then previous to that: "Knitting
Boar"
http://www.slideshare.net/jpatanooga/knitting-boar-atlhugjan2013v2

**Comment by crockpotveggies** [comment]
Josh, there are very valid points you've brought up. Adding complexity to an engineering problem is least desirable. 

**Re #2:** 
I think you'll find in the document in my original comment that I've referenced the Carnegie paper (Alex Smola's work) quite frequently. Much of this proposal is based on his work.

**Re the "use case":**
Here at Bernie A.I. we have our own personal cluster of 7 physical machines. We've got 56 Xeon cores to play with, and even though the hardware is 5+ years old, most of my research has told me that implementation (not hardware) achieves 50% of performance gains. 

Right now the current Spark implementation is a crawl. On my personal Macbook Air i7 processor, I can achieve 78% scoring in around 20-30 minutes using a small dataset. On our cluster, it's 24 hours plus. Even though iteration time on the server takes 30 seconds, the averaging step is a 15-30 minute process. There's a serious problem relying on the synchronicity of Spark. The use case of *some* modification is valid, whether it's a parameter server or serious makeover.

Some thoughts:
+ Are there any existing parameter server implementations that we can "drop in"?
+ How can we re-think the Spark implementation as is without a parameter server?
+ What if we re-thought training instead as a streaming job instead of a batch job? Replace MapReduce with Streaming architecture while still using Spark?

**Comment by crockpotveggies** [comment]
Actually, I want to revisit my last thought on my previous comment. Is Spark Streaming viable here? @AlexDBlack @agibsonccc 

We could potentially re-think training as a streaming process that fits within Spark Streaming:
1. There is a training data source that can be streamed: `val dataIterator = new EarlyStoppingDataIteratorSingleton(computationGraphConf)`
2. Training is an iterative process, and Spark allows for micro-batching: `val resultsGradient = data.map(new GradientFromTupleFunction())`
3. Gradients are then piped back into the Singleton for averaging while a second iteration is streamed: `dataIterator.setGradients(resultsGradient)`

The process is, of course, more complex than the pseudo-code I presented here. However, it's a talking point.

**Comment by jpatanooga** [comment]
On Fri, May 27, 2016 at 12:29 PM, crockpotveggies <notifications@github.com>
wrote:

> Josh, there are very valid points you've brought up. Adding complexity to
> an engineering problem is least desirable.
>
> *Re #2 <https://github.com/deeplearning4j/deeplearning4j/issues/2>:*
> I think you'll find in the document in my original comment that I've
> referenced the Carnegie paper (Alex Smola's work) quite frequently. Much of
> this proposal is based on his work.
>
> *Re the "use case":*
> Here at Bernie A.I. we have our own personal cluster of 7 physical
> machines. We've got 56 Xeon cores to play with, and even though the
> hardware is 5+ years old, most of my research has told me that
> implementation (not hardware) achieves 50% of performance gains.
>
> Right now the current Spark implementation is a crawl. On my personal
> Macbook Air i7 processor, I can achieve 78% scoring in around 20-30 minutes
> using a small dataset. On our cluster, it's 24 hours plus. Even though
> iteration time on the server takes 30 seconds, the averaging step is a
> 15-30 minute process. There's a serious problem relying on the
> synchronicity of Spark. The use case of *some* modification is valid,
> whether it's a parameter server or serious makeover.
>

Sounds like some profiling of the averaging step is in order to study where
the true bottlenecks are. From my own experiments I've found controlling
the batch of records that is partitioned per worker is a good way to
improve performance (read: dont pull massive batches into memory all at
once).

The other thing to consider from a distributed systems standpoint is that
sending lots of small network updates back and forth can degrade
hardware/network performance. Hardware and networks like operations to be
performed, in general, in large batches. So that's going to be a running
concern around how the updates were applied.


> Some thoughts:
>
>    - Are there any existing parameter server implementations that we can
>    "drop in"?
>    - How can we re-think the Spark implementation as is without a
>    parameter server?
>    - What if we re-thought training instead as a streaming job instead of
>    a batch job? Replace MapReduce with Streaming architecture while still
>    using Spark?
>
> - existing: most likely not, but not entirely sure.
- spark: we need to do some serious profiling around where all the time is
spent in spark. I've found that we have some improvements to go in terms of
how we're dealing w how much data is loaded into memory at once per worker.
The trick with spark, as opposed to something more raw (e.g. MapReduce), is
that you don't always have more visibility into what is going on (its
abstracted away).

- streaming: im not quite sure what you are proposing here. Parallel
iterative class algorithms are inherently multi-pass (epoch) over a
dataset. You can partition it different ways, and you can make the batch
reset time faster (e.g. the advantage of spark over MapReduce), but "using
streaming" is not something from my experience that would help. We're not
using MapReduce here, although from the perspective of a MapReduce job
being a single pass (epoch) across the dataset it might look similar ---
spark operates its job/DAG/task model differently so it can reset the pass
across the dataset faster. Similar concepts, variations in how they execute.

TLDR:

- let's focus on whats "not fast" today in spark and profile that
- once we know where those issues are, then we can have a discussion around
solving those specific issues (and simpler is better here)

**Comment by agibsonccc** [comment]
This is implemented now. Please move anything specific to the new parameter server implementation under nd4j.
